core/src/main/scala/org/http4s/internal/package.scala

```diff
@@ -14,37 +14,40 @@ import java.util.concurrent.{
 }
 
 import cats.effect.implicits._
-import cats.effect.{Async, Concurrent, ConcurrentEffect, ContextShift, Effect, IO}
+import cats.effect.std.Dispatcher
+import cats.effect.kernel.{Async, Sync}
 import cats.implicits._
 import fs2.{Chunk, Pipe, Pull, RaiseThrowable, Stream}
 import java.nio.{ByteBuffer, CharBuffer}
-import org.http4s.util.execution.direct
 import org.log4s.Logger
 
 import scala.concurrent.{ExecutionContext, Future}
 import scala.util.control.NoStackTrace
-import scala.util.{Failure, Success}
 import java.nio.charset.MalformedInputException
 import java.nio.charset.UnmappableCharacterException
 
 package object internal {
   // Like fs2.async.unsafeRunAsync before 1.0.  Convenient for when we
   // have an ExecutionContext but not a Timer.
-  private[http4s] def unsafeRunAsync[F[_], A](fa: F[A])(
-      f: Either[Throwable, A] => IO[Unit])(implicit F: Effect[F], ec: ExecutionContext): Unit =
-    F.runAsync(Async.shift(ec) *> fa)(f).unsafeRunSync()
+  private[http4s] def unsafeRunAsync[F[_]: Async, A](fa: F[A])(
+      f: Either[Throwable, A] => F[Unit])(implicit
+      dispatcher: Dispatcher[F],
+      ec: ExecutionContext): Unit =
+    dispatcher.unsafeRunSync(fa.evalOn(ec).attemptTap(f).void)
```

Comment:
you made this synchronous, your probably want to match the semantics of the existing code instead, also I'd be curious to see where this is used

Reply:
All the usages are in blaze.  We could punt it to blaze-core right now, and see if we can eliminate entirely when we get to that backend.

Reply:
What would be the right action for this PR? should I move it to blaze-core but under the same package, keep it, or delete it?

Reply:
Let's move it to blaze-core, under its natural package.  That will help get core clean so other ports can commence, and give me more time to relearn and answer to @djspiewak's question. :smile: